### ISSUE TITLE
no-unused-vars bug (for plus plus)

### ISSUE THREAD

**Issue by hezedu** [eslint/eslint#7578]
config is `"no-unused-vars": "warn"`
warn: 
```javascript
var j = 0; // j is assigned a value but never used
while(typeof window === 'object'){
  j++;
}
```
ok: 
```javascript
var j = 0;
while(typeof window === 'object'){
  j = j + 1;
}
```
can you make the  j++ to ok ?

**Comment by eslintbot** [comment]
Hi @hezedu, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by ilyavolodin** [comment]
@hezedu The rule is correctly flagging `j` as unused, because, it is, in fact, not used for anything. Value of `j` is not read, passed to a function, or consumed in any way, so the variable is unnecessary and can be removed without affecting result of the code.

**Comment by hezedu** [comment]
version 3.9.1

rule config none

expected to `j++` can be used

**Comment by platinumazure** [comment]
Look, the problem is that except for the increment, no one else sees the value of `j` at all. You could remove `j` entirely from your code and it would have the same effects to calling code.

Your code, in terms of observable side effects, is exactly the same as this code:

```js
while(typeof window === 'object'){
}
```

You need to finish writing your code (as in, write the lines where `j` is actually passed to another function or returned or used in an expression). Right now, the rule is correctly noting that `j` is not used except for self-modification.

**Comment by hezedu** [comment]
@platinumazure OK, thank, you are right.  and the `j = j + 1` not hint is a bug now?

**Comment by platinumazure** [comment]
Correct, it's not a bug. Even adding `console.log(j);` to your code should get rid of the lint warning, because that counts as a use of `j`.

**Comment by hezedu** [comment]
@platinumazure This problem is: same code,  `j++`  is lint,  `j = j + 1`  is not lint.

**Comment by platinumazure** [comment]
Ah, then `j = j + 1` should probably be flagged as well. Thanks for
clarifying, I'd say this is a bug.

**Comment by not-an-aardvark** [comment]
Looking at this again, I don't think it's a bug. In `j = j + 1`, the value of `j` is used in an expression, whereas `j++` always just increments `j`. Although the behavior happens to be the same in this case, that wouldn't be true for arbitrary expressions involving `j`, and I'm not sure it's a good idea to special-case expressions that typically don't have side-effects.

**Comment by vitorbal** [comment]
I agree with @not-an-aardvark, I think we should err on the side of caution here instead of adding a special case handling for this.

**Comment by hezedu** [comment]
@not-an-aardvark But the following simple example is still get a warn:
```js
var j = 0;
j = j + 1;
```

**Comment by not-an-aardvark** [comment]
Closing because this is working as intended.
